/*
 * Copyright (c) 2017.
 * THE WORK (AS DEFINED BELOW) IS PROVIDED UNDER THE TERMS OF THIS CREATIVE COMMONS PUBLIC LICENSE ("CCPL" OR "LICENSE"). THE WORK IS PROTECTED BY COPYRIGHT AND/OR OTHER APPLICABLE LAW. ANY USE OF THE WORK OTHER THAN AS AUTHORIZED UNDER THIS LICENSE OR COPYRIGHT LAW IS PROHIBITED.
 *
 * BY EXERCISING ANY RIGHTS TO THE WORK PROVIDED HERE, YOU ACCEPT AND AGREE TO BE BOUND BY THE TERMS OF THIS LICENSE. THE LICENSOR GRANTS YOU THE RIGHTS CONTAINED HERE IN CONSIDERATION OF YOUR ACCEPTANCE OF SUCH TERMS AND CONDITIONS.
 *
 * 1. Definitions
 *
 * "Collective Work" means a work, such as a periodical issue, anthology or encyclopedia, in which the Work in its entirety in unmodified form, along with a number of other contributions, constituting separate and independent works in themselves, are assembled into a collective whole. A work that constitutes a Collective Work will not be considered a Derivative Work (as defined below) for the purposes of this License.
 * "Derivative Work" means a work based upon the Work or upon the Work and other pre-existing works, such as a translation, musical arrangement, dramatization, fictionalization, motion picture version, sound recording, art reproduction, abridgment, condensation, or any other form in which the Work may be recast, transformed, or adapted, except that a work that constitutes a Collective Work will not be considered a Derivative Work for the purpose of this License. For the avoidance of doubt, where the Work is a musical composition or sound recording, the synchronization of the Work in timed-relation with a moving image ("synching") will be considered a Derivative Work for the purpose of this License.
 * "Licensor" means the individual or entity that offers the Work under the terms of this License.
 * "Original Author" means the individual or entity who created the Work.
 * "Work" means the copyrightable work of authorship offered under the terms of this License.
 * "You" means an individual or entity exercising rights under this License who has not previously violated the terms of this License with respect to the Work, or who has received express permission from the Licensor to exercise rights under this License despite a previous violation.
 * 2. Fair Use Rights. Nothing in this license is intended to reduce, limit, or restrict any rights arising from fair use, first sale or other limitations on the exclusive rights of the copyright owner under copyright law or other applicable laws.
 *
 * 3. License Grant. Subject to the terms and conditions of this License, Licensor hereby grants You a worldwide, royalty-free, non-exclusive, perpetual (for the duration of the applicable copyright) license to exercise the rights in the Work as stated below:
 *
 * to reproduce the Work, to incorporate the Work into one or more Collective Works, and to reproduce the Work as incorporated in the Collective Works;
 * to distribute copies or phonorecords of, display publicly, perform publicly, and perform publicly by means of a digital audio transmission the Work including as incorporated in Collective Works;
 * The above rights may be exercised in all media and formats whether now known or hereafter devised. The above rights include the right to make such modifications as are technically necessary to exercise the rights in other media and formats, but otherwise you have no rights to make Derivative Works. All rights not expressly granted by Licensor are hereby reserved, including but not limited to the rights set forth in Sections 4(d) and 4(e).
 *
 * 4. Restrictions.The license granted in Section 3 above is expressly made subject to and limited by the following restrictions:
 *
 * You may distribute, publicly display, publicly perform, or publicly digitally perform the Work only under the terms of this License, and You must include a copy of, or the Uniform Resource Identifier for, this License with every copy or phonorecord of the Work You distribute, publicly display, publicly perform, or publicly digitally perform. You may not offer or impose any terms on the Work that alter or restrict the terms of this License or the recipients' exercise of the rights granted hereunder. You may not sublicense the Work. You must keep intact all notices that refer to this License and to the disclaimer of warranties. You may not distribute, publicly display, publicly perform, or publicly digitally perform the Work with any technological measures that control access or use of the Work in a manner inconsistent with the terms of this License Agreement. The above applies to the Work as incorporated in a Collective Work, but this does not require the Collective Work apart from the Work itself to be made subject to the terms of this License. If You create a Collective Work, upon notice from any Licensor You must, to the extent practicable, remove from the Collective Work any reference to such Licensor or the Original Author, as requested.
 * You may not exercise any of the rights granted to You in Section 3 above in any manner that is primarily intended for or directed toward commercial advantage or private monetary compensation. The exchange of the Work for other copyrighted works by means of digital file-sharing or otherwise shall not be considered to be intended for or directed toward commercial advantage or private monetary compensation, provided there is no payment of any monetary compensation in connection with the exchange of copyrighted works.
 * If you distribute, publicly display, publicly perform, or publicly digitally perform the Work, You must keep intact all copyright notices for the Work and give the Original Author credit reasonable to the medium or means You are utilizing by conveying the name (or pseudonym if applicable) of the Original Author if supplied; the title of the Work if supplied; and to the extent reasonably practicable, the Uniform Resource Identifier, if any, that Licensor specifies to be associated with the Work, unless such URI does not refer to the copyright notice or licensing information for the Work. Such credit may be implemented in any reasonable manner; provided, however, that in the case of a Collective Work, at a minimum such credit will appear where any other comparable authorship credit appears and in a manner at least as prominent as such other comparable authorship credit.
 * For the avoidance of doubt, where the Work is a musical composition:
 *
 * Performance Royalties Under Blanket Licenses. Licensor reserves the exclusive right to collect, whether individually or via a performance rights society (e.g. ASCAP, BMI, SESAC), royalties for the public performance or public digital performance (e.g. webcast) of the Work if that performance is primarily intended for or directed toward commercial advantage or private monetary compensation.
 * Mechanical Rights and Statutory Royalties. Licensor reserves the exclusive right to collect, whether individually or via a music rights agency or designated agent (e.g. Harry Fox Agency), royalties for any phonorecord You create from the Work ("cover version") and distribute, subject to the compulsory license created by 17 USC Section 115 of the US Copyright Act (or the equivalent in other jurisdictions), if Your distribution of such cover version is primarily intended for or directed toward commercial advantage or private monetary compensation.
 * Webcasting Rights and Statutory Royalties. For the avoidance of doubt, where the Work is a sound recording, Licensor reserves the exclusive right to collect, whether individually or via a performance-rights society (e.g. SoundExchange), royalties for the public digital performance (e.g. webcast) of the Work, subject to the compulsory license created by 17 USC Section 114 of the US Copyright Act (or the equivalent in other jurisdictions), if Your public digital performance is primarily intended for or directed toward commercial advantage or private monetary compensation.
 * 5. Representations, Warranties and Disclaimer
 *
 * UNLESS OTHERWISE MUTUALLY AGREED BY THE PARTIES IN WRITING, LICENSOR OFFERS THE WORK AS-IS AND MAKES NO REPRESENTATIONS OR WARRANTIES OF ANY KIND CONCERNING THE WORK, EXPRESS, IMPLIED, STATUTORY OR OTHERWISE, INCLUDING, WITHOUT LIMITATION, WARRANTIES OF TITLE, MERCHANTIBILITY, FITNESS FOR A PARTICULAR PURPOSE, NONINFRINGEMENT, OR THE ABSENCE OF LATENT OR OTHER DEFECTS, ACCURACY, OR THE PRESENCE OF ABSENCE OF ERRORS, WHETHER OR NOT DISCOVERABLE. SOME JURISDICTIONS DO NOT ALLOW THE EXCLUSION OF IMPLIED WARRANTIES, SO SUCH EXCLUSION MAY NOT APPLY TO YOU.
 *
 * 6. Limitation on Liability. EXCEPT TO THE EXTENT REQUIRED BY APPLICABLE LAW, IN NO EVENT WILL LICENSOR BE LIABLE TO YOU ON ANY LEGAL THEORY FOR ANY SPECIAL, INCIDENTAL, CONSEQUENTIAL, PUNITIVE OR EXEMPLARY DAMAGES ARISING OUT OF THIS LICENSE OR THE USE OF THE WORK, EVEN IF LICENSOR HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH DAMAGES.
 *
 * 7. Termination
 *
 * This License and the rights granted hereunder will terminate automatically upon any breach by You of the terms of this License. Individuals or entities who have received Collective Works from You under this License, however, will not have their licenses terminated provided such individuals or entities remain in full compliance with those licenses. Sections 1, 2, 5, 6, 7, and 8 will survive any termination of this License.
 * Subject to the above terms and conditions, the license granted here is perpetual (for the duration of the applicable copyright in the Work). Notwithstanding the above, Licensor reserves the right to release the Work under different license terms or to stop distributing the Work at any time; provided, however that any such election will not serve to withdraw this License (or any other license that has been, or is required to be, granted under the terms of this License), and this License will continue in full force and effect unless terminated as stated above.
 * 8. Miscellaneous
 *
 * Each time You distribute or publicly digitally perform the Work or a Collective Work, the Licensor offers to the recipient a license to the Work on the same terms and conditions as the license granted to You under this License.
 * If any provision of this License is invalid or unenforceable under applicable law, it shall not affect the validity or enforceability of the remainder of the terms of this License, and without further action by the parties to this agreement, such provision shall be reformed to the minimum extent necessary to make such provision valid and enforceable.
 * No term or provision of this License shall be deemed waived and no breach consented to unless such waiver or consent shall be in writing and signed by the party to be charged with such waiver or consent.
 * This License constitutes the entire agreement between the parties with respect to the Work licensed here. There are no understandings, agreements or representations with respect to the Work not specified here. Licensor shall not be bound by any additional provisions that may appear in any communication from You. This License may not be modified without the mutual written agreement of the Licensor and You.
 *
 */

package com.csc;

import com.handlers.ArgsHandler;
import com.handlers.BamHandler;
import com.handlers.FisherStrandHandler;
import com.handlers.VCFHandler;

import java.io.*;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.logging.Level;
import java.util.logging.Logger;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * project checkStatusConcordance
 * Created by ayyoub on 6/8/17.
 */
class ConcordanceAnalysis {

    private HashMap<String, GenotypeStatus> hashQTLstatus = new HashMap<>();
    private HashMap<Integer, String> hashAniPosInVCF = new HashMap<>();
    private HashMap<String, HashMap<String, HashMap<String, GenotypeStatus>>> hashGenotype = new HashMap<>();
    private HashMap<String, HashMap<String, AllelicInfo>> alleleFrequency = new HashMap<>();

    private ArrayList<String> allAllele = new ArrayList<>();

    private ArrayList<List<Double>> fisherStrand;

    private String statusType;

    private int snpCount;


    ConcordanceAnalysis() {
        this.snpCount = 0;
    }


    /***********************************************************
     * ** check if names on VCF file and QTL status file match
     * @throws IOException file not found
     */
    private void checkNames()throws IOException {
        BufferedReader br = new BufferedReader(new FileReader(ArgsHandler.INSTANCE.getVcfPath()));
        String line;
        List<String> vcfIndiv = new ArrayList<>(10);
        List<String> qtlIndiv = new ArrayList<>(10);
        //Logger.getLogger("LOG").log(Level.INFO, "Checking homozygosity from VCF file");
        // store individuals from vcf file
        while ((line = br.readLine()) != null) {
            line = line.trim();
            if (line.startsWith("##"))
                continue;
            if (line.startsWith("#CHROM")) {
                String[] info = line.split("\t");
                vcfIndiv.addAll(Arrays.asList(info).subList(9, info.length));
                break;
            }
        }
        br.close();

        br = new BufferedReader(new FileReader(ArgsHandler.INSTANCE.getQtlPath()));
        // store individuals from qtl status file
        while ((line = br.readLine()) != null) {
            line = line.trim();
            qtlIndiv.add(line.split("\t")[0]);
        }
        br.close();

        // compare list of individuals
        if (vcfIndiv.size() != qtlIndiv.size()) {
            System.out.println("Number of individuals from VCF file doesn't match the QTL status file !");
            System.exit(0);
        }

        if (! vcfIndiv.containsAll(qtlIndiv)) {
            System.out.println("Individuals from VCF file doesn't match the ones from QTL status file !");
            System.exit(0);
        }

        if (! qtlIndiv.containsAll(vcfIndiv)) {
            System.out.println("Individuals from QTL status file doesn't match the ones from VCF file !");
            System.exit(0);
        }

    }


    /****************************************************
     * ** put genotype in a hash and check homozygosity
     * @throws IOException file not found
     */
    private void checkQTL() throws IOException, BadFileFormatException{
        BufferedReader br = new BufferedReader(new FileReader(ArgsHandler.INSTANCE.getQtlPath()));
        String line;
        System.out.println("Checking homozygosity from QTL status file");
        while ((line = br.readLine()) != null) {
            line = line.trim();
            String[] info = line.split("\t");
            String[] allele = info[1].split("");
            hashQTLstatus.put(info[0], new GenotypeStatus(info[1], allele[0], allele[1]));

            allAllele.add(allele[0]);
            allAllele.add(allele[1]);
        }

        br.close();

        if (allAllele.contains("P") && allAllele.contains("N")) {
            statusType = "PN";
        } else if (allAllele.contains("P")) {
            statusType = "P";
        } else if (allAllele.contains("N")) {
            statusType = "N";
        }


        for (String ani : hashQTLstatus.keySet()) {
            switch (statusType) {
                case "PN":
                    if (hashQTLstatus.get(ani).getGeno().matches("(.*)p(.*)|(.*)n(.*)|(.*)0(.*)|(.*)na(.*)")) {
                        hashQTLstatus.get(ani).setStatus("NA");
                    } else if (hashQTLstatus.get(ani).getGeno().equals("PP")) {
                        hashQTLstatus.get(ani).setStatus("homoP");
                    } else if (hashQTLstatus.get(ani).getGeno().equals("NN")) {
                        hashQTLstatus.get(ani).setStatus("homoN");
                    } else if (hashQTLstatus.get(ani).getGeno().equals("PN") || hashQTLstatus.get(ani).getGeno().equals("NP")
                            || hashQTLstatus.get(ani).getGeno().equals("Pn") || hashQTLstatus.get(ani).getGeno().equals("nP")
                            || hashQTLstatus.get(ani).getGeno().equals("P0") || hashQTLstatus.get(ani).getGeno().equals("0P")
                            || hashQTLstatus.get(ani).getGeno().equals("Np") || hashQTLstatus.get(ani).getGeno().equals("pN")
                            || hashQTLstatus.get(ani).getGeno().equals("N0") || hashQTLstatus.get(ani).getGeno().equals("0N")
                            || hashQTLstatus.get(ani).getGeno().equals("Pp") || hashQTLstatus.get(ani).getGeno().equals("pP")
                            || hashQTLstatus.get(ani).getGeno().equals("Nn") || hashQTLstatus.get(ani).getGeno().equals("nN")) {
                        hashQTLstatus.get(ani).setStatus("heterozygous");
                    }
                    break;
                case "P":
                    if (hashQTLstatus.get(ani).getGeno().matches("(.*)na(.*)")) {
                        hashQTLstatus.get(ani).setStatus("NA");
                    } else if (hashQTLstatus.get(ani).getGeno().equals("PP")) {
                        hashQTLstatus.get(ani).setStatus("homoP");
                    } else if (hashQTLstatus.get(ani).getGeno().equals("00") || hashQTLstatus.get(ani).getGeno().equals("0n")
                            || hashQTLstatus.get(ani).getGeno().equals("n0") || hashQTLstatus.get(ani).getGeno().equals("nn")
                            || hashQTLstatus.get(ani).getGeno().equals("pp") || hashQTLstatus.get(ani).getGeno().equals("p0")
                            || hashQTLstatus.get(ani).getGeno().equals("0p") || hashQTLstatus.get(ani).getGeno().equals("np")
                            || hashQTLstatus.get(ani).getGeno().equals("pn")) {
                        hashQTLstatus.get(ani).setStatus("homoN");
                    } else if (hashQTLstatus.get(ani).getGeno().equals("P0") || hashQTLstatus.get(ani).getGeno().equals("0P")
                            || hashQTLstatus.get(ani).getGeno().equals("Pn") || hashQTLstatus.get(ani).getGeno().equals("nP")
                            || hashQTLstatus.get(ani).getGeno().equals("Pp") || hashQTLstatus.get(ani).getGeno().equals("pP")) {
                        hashQTLstatus.get(ani).setStatus("heterozygous");
                    }
                    break;
                case "N":
                    if (hashQTLstatus.get(ani).getGeno().matches("(.*)na(.*)")) {
                        hashQTLstatus.get(ani).setStatus("NA");
                    } else if (hashQTLstatus.get(ani).getGeno().equals("NN")) {
                        hashQTLstatus.get(ani).setStatus("homoN");
                    } else if (hashQTLstatus.get(ani).getGeno().equals("00") || hashQTLstatus.get(ani).getGeno().equals("0n")
                            || hashQTLstatus.get(ani).getGeno().equals("n0") || hashQTLstatus.get(ani).getGeno().equals("nn")
                            || hashQTLstatus.get(ani).getGeno().equals("pp") || hashQTLstatus.get(ani).getGeno().equals("p0")
                            || hashQTLstatus.get(ani).getGeno().equals("0p") || hashQTLstatus.get(ani).getGeno().equals("np")
                            || hashQTLstatus.get(ani).getGeno().equals("pn")) {
                        hashQTLstatus.get(ani).setStatus("homoP");
                    } else if (hashQTLstatus.get(ani).getGeno().equals("N0") || hashQTLstatus.get(ani).getGeno().equals("0N")
                            || hashQTLstatus.get(ani).getGeno().equals("Nn") || hashQTLstatus.get(ani).getGeno().equals("nN")
                            || hashQTLstatus.get(ani).getGeno().equals("Np") || hashQTLstatus.get(ani).getGeno().equals("pN")) {
                        hashQTLstatus.get(ani).setStatus("heterozygous");
                    }
                    break;
            }

        }

    }

    /***************************************
     * @throws IOException file not found
     */
    private void checkGenotype() throws IOException {

        BufferedReader vcf = new BufferedReader(new FileReader(ArgsHandler.INSTANCE.getVcfPath()));
        String line;
        String chr;
        String pos;
        String alt;
        String ref;
        int gqIndex = -1;
        System.out.println("Checking homozygosity from VCF file");
        while ((line = vcf.readLine()) != null) {
            line = line.trim();
            if (line.startsWith("##"))
                continue;
            if (line.startsWith("#CHROM")) {
                String[] info = line.split("\t");
                // change this for the general case
                for (int i = 9; i < info.length; i++) {
                    hashAniPosInVCF.put(i, info[i]); // added for general case
                } // END of for i
            } // END of if #CHROM
            else {
                if (gqIndex == -1) {
                    String[] info = line.split("\t")[8].split(":");
                    gqIndex = Arrays.asList(info).indexOf("GQ");
                }
                snpCount++;
                String[] info = line.split("\t");
                if (!info[0].startsWith("Chr"))
                    continue;
                if (info[4].contains(","))
                    continue;
                chr = info[0].substring(3);
                pos = info[1];
                alt = info[4];
                ref = info[3];

                for (int i : hashAniPosInVCF.keySet()) {
                    int gq;
                    try {
                        gq = Integer.parseInt(info[i].split(":")[gqIndex]);
                    } catch (NumberFormatException e) {
                        continue;
                    }
                    // check if genotype quality >= 20
                    if (gq < ArgsHandler.INSTANCE.getGqThr()) {
                        continue;
                    }
                    // check fisher strand < threshold
                    if (fisherStrand != null) {
                        // i starts from 9
                        assert i - 9 >= 0;
                        if(fisherStrand.get(i - 9).get(snpCount - 1) < ArgsHandler.INSTANCE.getFsThr())
                            continue;
                    }

                    String geno = info[i];
                    Pattern p = Pattern.compile("^([01])/([01]):.*$");
                    Matcher m = p.matcher(geno);
                    if (m.find()) {
                        if (m.group(1).equals(m.group(2))) {
                            if (hashGenotype.get(chr) == null) {
                                hashGenotype.put(chr, new HashMap<>());
                                alleleFrequency.put(chr, new HashMap<>());
                            }

                            if (hashGenotype.get(chr).get(pos) == null) {
                                hashGenotype.get(chr).put(pos, new HashMap<>());
                                alleleFrequency.get(chr).put(pos, new AllelicInfo());
                            }

                            hashGenotype.get(chr).get(pos).put(hashAniPosInVCF.get(i),
                                    new GenotypeStatus("homozygous", m.group(1), m.group(2), alt, ref));
                            alleleFrequency.get(chr).get(pos).incrementAlleleCount(m.group(1), m.group(2));
                            alleleFrequency.get(chr).get(pos).incrementHomCount();

                        } else {
                            if (hashGenotype.get(chr) == null) {
                                hashGenotype.put(chr, new HashMap<>());
                                alleleFrequency.put(chr, new HashMap<>());
                            }

                            if (hashGenotype.get(chr).get(pos) == null) {
                                hashGenotype.get(chr).put(pos, new HashMap<>());
                                alleleFrequency.get(chr).put(pos, new AllelicInfo());
                            }

                            hashGenotype.get(chr).get(pos).put(hashAniPosInVCF.get(i),
                                    new GenotypeStatus("heterozygous", m.group(1), m.group(2), alt, ref));
                            alleleFrequency.get(chr).get(pos).incrementAlleleCount(m.group(1), m.group(2));
                            alleleFrequency.get(chr).get(pos).incrementHetCount();
                        }
                    } // END of if m.find
                } // END of for i
            }
        } // END of while vcf

        vcf.close();
    }

    /**************************************
     * ** compare genotype and QTL status
     *
     * @throws IOException ***
     */
    private void compareStatus() throws IOException {
        System.out.println("Comparing genotype and QTL status");
        // add extension
        BufferedWriter bw = new BufferedWriter(new FileWriter(ArgsHandler.INSTANCE.getOutPath() + MainEntry.EXTENSION));
        bw.write("chr\tposition\tnb_concordant\ttotal individuals\tperc_concordance\tconcordant by chance\tref\talt");
        bw.newLine();
        String alt = null;
        String ref = null;
        for (String chr : hashGenotype.keySet()) {
            for (String pos : hashGenotype.get(chr).keySet()) {
                int concordant1 = 0;
                int concordant2 = 0;
                int tot = 0;
                for (String ani : hashGenotype.get(chr).get(pos).keySet()) {
                    alt = hashGenotype.get(chr).get(pos).get(ani).getAlt();
                    ref = hashGenotype.get(chr).get(pos).get(ani).getRef();
                    if (hashQTLstatus.get(ani).getStatus().equals("NA"))
                        continue;

                    if (hashGenotype.get(chr).get(pos).get(ani).getGeno().equals("heterozygous")) {
                        if (hashQTLstatus.get(ani).getStatus().equals("heterozygous")) {
                            concordant1++;
                            concordant2++;
                            tot++;
                            continue;
                        } else {
                            tot++;
                            continue;
                        }
                    }

                    if (hashGenotype.get(chr).get(pos).get(ani).getGeno().equals("homozygous")) {
                        if (hashQTLstatus.get(ani).getStatus().equals("homoP")) {
                            if (hashGenotype.get(chr).get(pos).get(ani).getAllele1().equals("0"))
                                concordant1++;
                            if (hashGenotype.get(chr).get(pos).get(ani).getAllele1().equals("1"))
                                concordant2++;
                            tot++;
                        } else if (hashQTLstatus.get(ani).getStatus().equals("homoN")) {
                            if (hashGenotype.get(chr).get(pos).get(ani).getAllele1().equals("1"))
                                concordant1++;
                            if (hashGenotype.get(chr).get(pos).get(ani).getAllele1().equals("0"))
                                concordant2++;
                            tot++;
                        } else {
                            tot++;
                        }
                    }
                } // END of for ani
                int concordant = concordant1 >= concordant2 ? concordant1 : concordant2;

                StringBuilder buffer = new StringBuilder();
                buffer.append(chr).append("\t");
                buffer.append(pos).append("\t");
                buffer.append(concordant).append("\t");
                buffer.append(tot).append("\t");
                if (tot != 0)
                    buffer.append(String.format("%.2f", ((double) concordant / (double) tot)));
                bw.write(buffer.toString());
                bw.write("\t");
                double p = alleleFrequency.get(chr).get(pos).getRefAlleleFrequency();
                int n = alleleFrequency.get(chr).get(pos).getHetCount();
                int m = alleleFrequency.get(chr).get(pos).getHomCount();
                double cbc = 2 * Math.pow(p * (1 - p), n) * Math.pow((1 - 2 * p * (1 - p)), m);
                if (cbc <= (1 / snpCount))
                    bw.write("NO\t");
                else
                    bw.write("YES\t");
                assert ref != null;
                bw.write(ref);
                bw.write("\t");
                bw.write(alt);
                bw.newLine();

            } // END of for pos


        } // END of for chr


        bw.close();
    }


    void run() {
        // call checkNames
        try {
            checkNames();
        } catch (IOException e) {
            e.printStackTrace();
        }

        // check if BAM file and ref file are provided, else don't compute fisher strand
        if (ArgsHandler.INSTANCE.getBamFiles() != null && ArgsHandler.INSTANCE.getRef() !=  null) {
            // call vcfHandler
            VCFHandler vcfHandler = new VCFHandler(ArgsHandler.INSTANCE.getVcfPath());
            // returns the path of the file containing positions, should i use it (PS: hard coded in BamHandler)?
            vcfHandler.extractPositions();
            // call bamHamdler
            BamHandler bamHandler = new BamHandler();
            try {
                bamHandler.extractSamplesNames(ArgsHandler.INSTANCE.getBamFiles());
            } catch (ReadGroupException e1){
                System.out.println("Please add read group information to BAM files or skip fisher strand computation.");
                System.exit(3);
            } catch (BadFileFormatException e2){
                System.out.println("Please check BAM files integrity or skip fisher strand computation.");
            }

            bamHandler.mpileup(ArgsHandler.INSTANCE.getBamFiles(), ArgsHandler.INSTANCE.getRef());
            // call fisherStrandHandler
            FisherStrandHandler fsHandler = new FisherStrandHandler();
            fisherStrand = new ArrayList<>(ArgsHandler.INSTANCE.getBamFiles().size());
            for (int i = 0; i < ArgsHandler.INSTANCE.getBamFiles().size(); i++) {
                fisherStrand.add(fsHandler.countReads(i));
            }
        } else {
            System.out.println("BAM files or reference file not provided! Skipping fisher strand computation.");
        }


        try {
            checkQTL();
        } catch (IOException e) {
            e.printStackTrace();
        }
        try {
            checkGenotype();
        } catch (IOException e) {
            e.printStackTrace();
        }
        try {
            compareStatus();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}
